### ISSUE TITLE
Band Structure 

### ISSUE THREAD

**Issue by cortner** [cortner/SKTB.jl#8]
* [ ] need a robust fermi-level solver that can deal with large beta
* [ ] Combine the two band structure functions into a single one, with a choice of energy interval. Low priority for me, but could address at some point.
* [x] can we add some functionality for some simple band structure diagrams? i.e. plotting along paths in BZ?

**Comment by jarvist** [comment]
Being able to plot actual band structures would be very useful.

There are standard resources to generate the list of high symmetry locations in the BZ to go between. 
The standard tabulation is [Bradley and Cracknell](https://books.google.co.uk/books?isbn=0199582580), a more modern method is via [seekpath](https://www.materialscloud.org/work/tools/seekpath). There's no obvious need to reproduce this work. 

From these systems, you end up with a file that often looks something like:
```
80  0 0 0 .666666667 0 0 G to K
```
Here 80 is the requested number of data points; (0,0,0) is the Gamma point, (2/3,0,0) is the K point (in this setting). It's easy enough to generate a list of k-points, and perhaps is best left to the user.

However, you do need some way to inject this set of custom k-points into the TB Hamiltonian. 

A function
`bandsatkpoint(kpoint,tbm)` 
would be a very useful starting point! 

I guess this will have to wrap `full!`/'sparse!': (https://github.com/cortner/TightBinding.jl/blob/177a98c8bc4d883c95249c93ce7efd89498ef808/src/matrixassembly.jl#L247-L265) 

It would probably be most easy to specify the k-points in BZ units [0..1], rather than bother to convert them to reciprocal lattice units, only to then convert them back again when constructing the phase-factor.

**Comment by cortner** [comment]
how does #21 look?

**Comment by jarvist** [comment]
Looks fantastic!

**Comment by jarvist** [comment]
I hacked together a bit of code to get ~sensible-ish band structures for high symmetry lines. 
Do you think this is worth formalising + making part of the code? 
This high-symmetry route (for `:hcp`) is just by hand, but there's a few other generic ones we could add, as well as exposing a way of specifying by hand. 

![image](https://user-images.githubusercontent.com/82852/39867691-0953ed6a-544e-11e8-9023-97c41ba871b7.png)

```
# This is a standard route in reciprocal space for hexagonal cells
spec=
"100  0 0 0 .666666667 0 0 G to K
100  .666666667 0 0 .5 -.288675 0 K to M
100  .5 -.288675 0 0 0 0 M to G
100  0 0 0 0 0 .306186 G to A
100  .666666667 0 0 .666666667 0 .306186  K to H
100  .666666667 0 .306186 .5 -.288675 .306186 H to L
100  .5 -.288675 .306186 0 0 .306186 L to A
100  0 0 .306186 .666666667 0 .306186 A to H"

rows(M::Matrix) = map(x->reshape(getindex(M, x, :), :, size(M)[2]), 1:size(M)[1])
columns(M::Matrix) = map(x->getindex(M, :, x), 1:size(M)[2])

r=readdlm(IOBuffer(spec))

const BZstep=0.005 # step in reciprocal space units

K=[]
points=[0]
labels=[]
point=0

for l in rows(r)
    println("line: $l ")
    N=l[1]
    kxa=l[2]; kya=l[3]; kza=l[4];
    kxb=l[5]; kyb=l[6]; kzb=l[7];

    # vector form
    ka=[kxa kya kza]
    kb=[kxb kyb kzb]
    println("  ka: $ka kb: $kb")
    
    N=ceil(norm((kb-ka)/BZstep)) # Choose number of data points by size in BZ
    step=(kb-ka)/AbstractFloat(N)
    println("N: $N  step: $step")

    
    for n=1:N
        k=ka + step*n
#        println("k: $k k[1]:",k[1])
        push!(K,JVecF(k[1],k[2],k[3]))
    end
   
    point+=N
    append!(points,point) # location of high symmetry point
#    label=l[8] - this is a bit rubbish, you almost always are going to need to write these yourself
#    append!(labels,label) # label of high symmetry point
end
```

```
# plot the resulting Band Structure
bands = TB.band_structure(K, nrltb, at)
gr() # v. compact SVG (?) files in notebook viewer, but static
#plotly() # nice for 'live' mouseover-able figures, but generates ENORMOUS ipython notebook save files, which don't generally work in the online viewer...

plot(bands, label="", xticks=(points,["G","K","M","G","A/K","H","L","A","H"]) )
hline!([eF],line=(1,:dash,:black),label="")
hline!([0],line=(1,:dot,:black),label="")

vline!(points, line=(0.5,:black),label="")
```

**Comment by cortner** [comment]
I would love for this to be part of `TightBinding.jl`. I assume you plan to 
* adding built-in "standard" high-symmetry lines for the most common crystal structures  (maybe bcc, fcc, hcp, diamond cubic, ???)
* have a simple way to generate paths by hand (e.g. specify corners)
* load paths from a file
* generate the figure

I think only the last one we should discuss for a moment, since it means that `TightBinding.jl` will need `Plots.jl` as a dependency. I'd be happy if we can make it an optional dependency, would that be ok for you?

**Comment by cortner** [comment]
btw, if you are really planning to continue to work on this code, and not just temporarily play around with it, then I'd be very happy to give you direct access to the repository. It would still be nice to have PRs for larger changes/enhancements, but bug-fixes and similar could just be checked directly into master. Would you prefer this?

**Comment by jarvist** [comment]
Definitely planning to continue working on these codes. Direct repository access would be useful - if only to assign tasks to myself! More than happy to more formally prepare PRs for larger changes. As we're not co-located, we may as well have our discussion out in the open via GitHub.

**Comment by cortner** [comment]
done!

**Comment by jarvist** [comment]
Your plan agrees with what I was thinking.

`Plots.jl` dependency is a problem. 
I think the recommended practice is that you should use https://github.com/JuliaPlots/RecipesBase.jl , but I haven't actually spend the time to figure out the magic metaprogramming that needs. 

For a previous package of mine, I sort of wrapped a plotting module-in-a-module, and have the dependency on Plots.jl undeclared, but it's a bit ugly as method.
https://github.com/jarvist/PolaronMobility.jl/blob/d3b46b9f8ba9df49e1f12b7d530c57f962aa776c/src/PlotPolaron.jl

**Comment by cortner** [comment]
Btw, I'll be at IC on May 31 for a PhD viva, at KCL on 19 July for the TYC Soiree.

**Comment by cortner** [comment]
> I think the recommended practice is that you should use https://github.com/JuliaPlots/RecipesBase.jl , but I haven't actually spend the time to figure out the magic metaprogramming that needs.

ok - so you could define a `struct BandDiagram` for which you could then specify a recipe.

**Comment by cortner** [comment]
It might actually be good to figure out those recipes, I should do this as well actually.

**Comment by cortner** [comment]
yeah - I just checked, 9s load time for `Plots.jl`, it would be better to not have this a dependency.
